lib/peep/prometheus.ex

```diff
@@ -53,89 +49,109 @@ defmodule Peep.Prometheus do
+      [name, "_count", labels_done, Integer.to_string(count + inf), ?\n]
+    ]
   end
 
   defp format_standard({metric, series}, type) do
     name = format_name(metric.name)
-    help = "# HELP #{name} #{escape_help(metric.description)}"
-    type = "# TYPE #{name} #{type}"
+    help = ["# HELP ", name, " ", escape_help(metric.description)]
+    type = ["# TYPE ", name, " ", to_string(type)]
 
     samples =
-      Enum.map_intersperse(series, ?\n, fn {labels, value} ->
+      Enum.map(series, fn {labels, value} ->
         has_labels? = not Enum.empty?(labels)
 
         if has_labels? do
-          "#{name}{#{format_labels(labels)}} #{format_value(value)}"
+          [name, ?{, format_labels(labels), ?}, " ", format_value(value), ?\n]
         else
-          "#{name} #{format_value(value)}"
+          [name, " ", format_value(value), ?\n]
         end
       end)
 
-    Enum.intersperse([help, type, samples], ?\n)
+    [help, ?\n, type, ?\n, samples]
   end
 
   defp format_labels(labels) do
     labels
-    |> Enum.map(fn {k, v} -> ~s/#{k}="#{escape(v)}"/ end)
     |> Enum.sort()
-    |> Enum.intersperse(",")
+    |> Enum.map_intersperse(?,, fn {k, v} -> [to_string(k), "=\"", escape(v), ?"] end)
   end
 
   defp format_name(name) do
     name
     |> Enum.join("_")
-    |> String.replace(~r/[^a-zA-Z0-9_]/, "")
-    |> String.replace(~r/^[^a-zA-Z]+/, "")
+    |> format_name_start()
   end
 
+  # Name must start with an ascii letter
+  defp format_name_start(<<h, rest::binary>>) when h not in ?A..?Z and h not in ?a..?z,
+    do: format_name_start(rest)
+
+  defp format_name_start(<<rest::binary>>),
+    do: format_name_rest(rest, <<>>)
+
+  # Otherwise only letters, numbers, or _
+  defp format_name_rest(<<h, rest::binary>>, acc)
+       when h in ?A..?Z or h in ?a..?z or h in ?0..?9 or h == ?_,
+       do: format_name_rest(rest, <<acc::binary, h>>)
+
+  defp format_name_rest(<<_, rest::binary>>, acc), do: format_name_rest(rest, acc)
+  defp format_name_rest(<<>>, acc), do: acc
+
   defp format_value(true), do: "1"
   defp format_value(false), do: "0"
   defp format_value(nil), do: "0"
-  defp format_value(n) when is_integer(n), do: :erlang.integer_to_binary(n)
-  defp format_value(f) when is_float(f), do: :erlang.float_to_binary(f, [:compact])
+  defp format_value(n) when is_integer(n), do: Integer.to_string(n)
+  defp format_value(f) when is_float(f), do: Float.to_string(f)
 
   defp escape(nil), do: "nil"
 
   defp escape(value) do
     value
     |> to_string()
-    |> String.replace(~S("), ~S(\"))
-    |> String.replace(~S(\\), ~S(\\\\))
-    |> String.replace(~S(\n), ~S(\\n))
+    |> escape(<<>>)
   end
 
+  defp escape(<<?\", rest::binary>>, acc), do: escape_help(rest, <<acc::binary, ?\\, ?\">>)
+  defp escape(<<?\\, rest::binary>>, acc), do: escape_help(rest, <<acc::binary, ?\\, ?\\>>)
+  defp escape(<<?\n, rest::binary>>, acc), do: escape_help(rest, <<acc::binary, ?\\, ?\n>>)
```

Comment:
We could use a guard but I chose to have multiple clauses for clarity, your call.